crates/noirc_evaluator/src/ssa/acir_gen/constraints.rs

```diff
@@ -405,9 +405,10 @@ pub(crate) fn to_radix_base(
     evaluator: &mut Evaluator,
 ) -> Vec<Witness> {
     // ensure there is no overflow
-    let mut max = BigUint::from(radix);
-    max = max.pow(limb_size) - BigUint::one();
-    assert!(max < FieldElement::modulus());
+    let mut min = BigUint::from(radix);
+    min = min.pow(limb_size - 1);
+    // minimum value that can be represented with limb_size limbs should be less than the modulus
+    assert!(min < FieldElement::modulus());
```

Comment:
Can you explain why this is valid?

Reply:
I'm not entirely sure this is the right approach, but here's my reasoning:

Ideally, `to_radix_base` should work for all elements in the range of $0 \leq x < p$, regardless of the `radix` and `limb_size` values, where $p$ represents the modulus.

Let's consider the case when `radix = 2` and $p = 13$. Under the previous stricter condition, a query with `limb_size = 4` would trigger an assertion failure, as $2^4 - 1 < 13$. Therefore, any 4-digit binary number (8 to 12) couldn't be accurately represented.

The new condition would accommodate queries with `limb_size = 4`, but would fail for `limb_size = 5`. While this condition is more lenient, it introduces the risk of overflow if the expression contains a value $\geq p$ (in this case, 13 to 15). I'm not sure whether this scenario is feasible, as the expression usually consists of a `FieldElement`, thereby implicitly enforcing a $< p$ range check. However, correct me if I'm wrong.

In the general bn-254 case, with radix = 2, the former condition would fail for `limb_size = 254`. This means that any number (less than the modulus) requiring full 254 bits could not be represented.

Reply:
I agree that the stricter overflow condition prevents some elements from being represented but it yields correct result, while weakening it does not ensure the unicity of the decomposition.
With your example of radix 2 and p=13, you can have 2 decomposed into (1,1,1,1).

It is possible to make the to_radix works properly for any field though, do you need this feature?

Reply:
Support for decomposing with `limb_size = 254` is required for https://github.com/noir-lang/noir/issues/1129

Reply:
> I agree that the stricter overflow condition prevents some elements from being represented but it yields correct result, while weakening it does not ensure the unicity of the decomposition. With your example of radix 2 and p=13, you can have 2 decomposed into (1,1,1,1).

Thanks for the explanation. I got your point now. In that case, I'll revert this change.

For the EdDSA use-case, `to_le_bits` is required for efficient scalar multiplication of an elliptic curve point. 
https://github.com/noir-lang/noir/blob/04f24e5f1518cfc039c0b36c68bf7e9807b0ad36/noir_stdlib/src/eddsa.nr#L28-L50

Since `hash` is a `FieldElement`, there's a chance that the MSB in the 254-bit representation is 1. Hence, you need full 254-bits when doing double-and-add multiplication:

https://github.com/noir-lang/noir/blob/04f24e5f1518cfc039c0b36c68bf7e9807b0ad36/noir_stdlib/src/ec/tecurve.nr#L371-L385

I guess a workaround, for now, could be to calculate the LSB in noir and only query the bit representation of the remaining bits.
  
```rust
// Scalar multiplication (p + ... + p n times)
fn mul(self, n: Field, p: Point) -> Point {
    let lsb = n as u1;
    let prefix = (n - lsb as Field) / 2;

    let bits = crate::field::modulus_num_bits() - 1;
    let prefix_as_bits = prefix.to_le_bits(bits as u32);

    let mut out = Point::zero();
    for i in 0..bits {
        out = self.add(
            self.add(out, out),
            if(prefix_as_bits[bits - i - 1] == 0) {Point::zero()} else {p});
    }
    out = self.add(
        self.add(out, out),
        if(lsb == 0) {Point::zero()} else {p});

    out
}
```

But this increases the number of constraints quite a bit. The [EdDSA test](https://github.com/noir-lang/noir/blob/04f24e5f1518cfc039c0b36c68bf7e9807b0ad36/noir_stdlib/src/eddsa.nr) goes from ~145k to ~247k constraints.

Reply:
I am not sure we can avoid the overhead, I will try to see if we get better results if it is done in noir.